internal/config/instance.go

```diff
@@ -113,65 +117,68 @@ func Get() (*Instance, error) {
 }
 
 // Set sets a value at the given key
-func (i *Instance) Set(key string, value interface{}) {
+func (i *Instance) Set(key string, value interface{}) error {
 	i.rwLock.Lock()
 	defer i.rwLock.Unlock()
-	i.viper.Set(key, value)
+
+	err := i.ReadInConfig()
+	if err != nil {
+		return err
+	}
+
+	i.data[strings.ToLower(key)] = value
+
+	err = i.save()
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// GetString retrieves a string for a given key
-func (i *Instance) GetString(key string) string {
+func (i *Instance) get(key string) interface{} {
 	i.rwLock.RLock()
 	defer i.rwLock.RUnlock()
-	return i.viper.GetString(key)
+	return i.data[strings.ToLower(key)]
 }
 
+// GetString retrieves a string for a given key
+func (i *Instance) GetString(key string) string {
+	return cast.ToString(i.get(key))
+}
+
+// AllKeys returns all of the curent config keys
 func (i *Instance) AllKeys() []string {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.AllKeys()
+	var keys []string
+	for k := range i.data {
+		keys = append(keys, k)
+	}
+	return keys
 }
 
 // GetStringMapStringSlice retrieves a map of string slices for a given key
 func (i *Instance) GetStringMapStringSlice(key string) map[string][]string {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.GetStringMapStringSlice(key)
-}
-
-// SetDefault sets the default value for a given key
-func (i *Instance) SetDefault(key string, value interface{}) {
-	i.rwLock.Lock()
-	defer i.rwLock.Unlock()
-	i.viper.SetDefault(key, value)
+	return cast.ToStringMapStringSlice(i.get(key))
 }
 
 // GetBool retrieves a boolean value for a given key
 func (i *Instance) GetBool(key string) bool {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.GetBool(key)
+	return cast.ToBool(i.get(key))
 }
 
 // GetStringSlice retrieves a slice of strings for a given key
 func (i *Instance) GetStringSlice(key string) []string {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.GetStringSlice(key)
+	return cast.ToStringSlice(i.get(key))
 }
 
 // GetTime retrieves a time instance for a given key
 func (i *Instance) GetTime(key string) time.Time {
-	i.rwLock.RLock()
-	defer i.rwLock.RUnlock()
-	return i.viper.GetTime(key)
+	return cast.ToTime(i.get(key))
```

Comment:
Another aside (may be worth thinking about now): The API surface can be reduced while increasing capability using this form: https://play.golang.org/p/sCtajzdwwTc

Many types can be added to the switch so that any reasonable values are handled, which is the compromise: On one hand, the methods can reveal what types are supported, but adding types adds noise to the API. On the other hand, many types can be added without any extra noise, but it is not fully clear what types are supported.

If this is interesting, but not for this story, please add a story or let me know and I can do so.

Reply:
This would also be another fairly drastic change to the config API. I'm not against it but do not think it's appropriate for this story.

Reply:
Story here: https://www.pivotaltracker.com/story/show/177194225. This also covers your error handling proposal.